Mathlib/Topology/Compactification/OnePointEquiv.lean

```diff
@@ -0,0 +1,166 @@
+
+## Main definitions
+
+* `divOnePoint`: Division on `K` taking values in `OnePoint K`. Notation: `÷`.
+* `divSlope`: Lowering `divOnePoint` to a map `ℙ K (Fin 2 → K) → OnePoint K`.
+
+## Main results
+
+* `divSlope_equiv` proves that `ℙ K (Fin 2 → K) ≃ OnePoint K`.
+
+## Tags
+
+one-point extension, projectivization
+-/
+
+open scoped LinearAlgebra.Projectivization OnePoint
+open Projectivization Classical
+
+/-!
+### Definition and basic properties
+
+We define a division operation `divOnePoint`, with notation `÷`.
+It takes inputs from a field `K` and outputs a value in `K ∪ {∞}`, i.e.,  `OnePoint K`.
+We prove that `÷` preserves projective equivalence: `(c•a)÷(c•b)=a÷b`.
+Hence `÷` projects down to a map `divSlope` from `ℙ K (Fin 2 → K)` to `OnePoint K`
+that takes a line through the origin and returns its slope. For a vertical line
+the slope is `∞`.
+-/
+variable {K : Type*}
+/-- A modified division from a `DivisionRing` to its `OnePoint` extension. With notation `÷`
+for `divOnePoint`, we have in particular that `1 ÷ 0 = ∞`. -/
+noncomputable def divOnePoint [DivisionRing K] (a : K) (r : K): OnePoint K :=
+  if r ≠ 0 then a / r else ∞
+
+/-- Uncurried version of `divOnePoint`, with nonzeroness assumption. -/
+noncomputable def divOnePoint' [DivisionRing K]
+  (u : {v : Fin 2 → K // v ≠ 0}) : OnePoint K :=
+  if u.1 1 ≠ 0 then some (u.1 0 / u.1 1) else ∞
+
+/-- Notation `÷` showing that `divOnePoint` is distinct from
+the ordinary division `/` of a `DivisionRing`. -/
+infix:50 " ÷ " => divOnePoint
+
+/-- `divOnePoint` can be lifted to the projective line (see `divSlope`.) -/
+lemma divOnePoint_lifts [Field K] (a b : {v : Fin 2 → K // v ≠ 0})
+    (h : ∃ c : Kˣ, (fun m : Kˣ ↦ m • b.1) c = a.1) :
+    (fun u ↦ u.1 0 ÷ u.1 1) a = (fun u ↦ u.1 0 ÷ u.1 1) b := by
+  obtain ⟨c,hc⟩ := h
+  simp_all only
+  rw [← hc]
+  unfold divOnePoint; simp only [ne_eq, Fin.isValue, Pi.smul_apply, ite_not]
+  split_ifs with hbc hb hb
+  · rfl
+  · simp_all only [ne_eq, OnePoint.infty_ne_coe]
+    exact hb <| (Units.mul_right_eq_zero c).mp hbc
+  · rw [hb] at hbc
+    simp at hbc
+  · apply congrArg some
+    field_simp
+    show c * b.1 0 * b.1 1 = b.1 0 * (c * b.1 1)
+    ring
+
+/-- Equivalence between the projective line and the one-point extension. -/
+noncomputable def divSlope [Field K] (p : ℙ K (Fin 2 → K)) : OnePoint K :=
+  Quotient.lift (fun u => divOnePoint (u.1 0) (u.1 1)) divOnePoint_lifts p
+
+/-! ### Equivalence
+We establish the equivalence between `OnePoint K` and `ℙ K (Fin 2 → K)` for a field `K`.
+-/
+
+/-- In a nonzero pair, if one coordinate is 0 then the other is nonzero. -/
+lemma not_both_zero [Zero K]
+    (a : Fin 2 → K) (ha : a ≠ 0) (h : a 1 = 0) : a 0 ≠ 0 := by
+  intro hc; apply ha; ext s
+  cases (Nat.le_one_iff_eq_zero_or_eq_one.mp (Fin.is_le s)) with
+  |inl hl => simp_all [Fin.eq_of_val_eq (j := 0) hl]
+  |inr hr => simp_all [Fin.eq_of_val_eq (j := 1) hr]
+
+instance [DivisionRing K] : Setoid ({v : Fin 2 → K // v ≠ 0}) :=
```

Comment:
I think it is better to make this instance `local`.

Reply:
Actually that instance is not needed so I just removed it.